### PR TITLE
feat(data): add docs for CRUD operations between Storage files and GraphQL API records

### PR DESCRIPTION
**Pull request by david-mcafee**
### Changes are hosted here: https://pr-5503.d39agkzf0cwtdz.amplifyapp.com/lib/graphqlapi/working-with-files/q/platform/js/

feat(data): add docs for CRUD operations between Storage files and GraphQL API records

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [x] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [x] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [x] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [x] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [x] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:3]: Let us remove the wording on complex here. I think that is confusing for customers, and makes it seem unattainable or difficult to implement.

Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx]: Just saw your comment, but I've already removed this. Thanks!

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:15]: For copy and paste-ability, let's just make sure the global auth rule is set or `@auth(rules: [{ allow: public }])`

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:24]: I think this callout needs to be worded stronger. Customers really need to understand that each category (storage + API) governs data access based on their own authZ patterns. We need to be explicit that "@auth rules" don't guard access in Storage / S3.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:26]: Love this.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:30]: > The corresponding file will be associated with this record.

In the code snippets below you don't talk about the file just yet. So I think there's a bit of a disconnect here for the reader.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:28]: For SEO, we should make these titles more explicit to "API + Storage". Feedback applies to all H2 headings below.

For example: "Create a record with an attached file"

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:32]: This example is maybe too minimal, as customer are able to get similar code snippets from other areas of our docs. I think we can expand the current section to "Create a record with an attached file". I.e. also show how to associate the file.

For the reader, we should answer the question: "What's Amplify's recommended guidance? Upload file first, then create record or vice versa?"

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:85]: Expand code snippets based on feedback above.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:124]: Feedback attached above. Let's combine the snippets between Storage and API workflows and not separate it out. One recurring theme from customer conversations and that came up in the Discord office hours is "What's the sequence of events I should do for Storage + API, which one goes first? Is there guidance?"

Comment (renebrandel) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:124]: Just want to make sure we answer that question "head-on"

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:173]: In the `meta.description`, the word "attachment" is used which is inconsistent with "associated". Let's just pick one and remain consistent. I prefer attached/attachment but not strongly opinionated.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:283]: Just to stick to the terminology above. This should probably be something like:
### 1. Disassociate / Detach file from the record

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:281]: I think this warrants a bit more expansion. To me it reads like "there are three primary deletion workflows to when working with files and GraphQL API."
1. Remove both record and attached files
2. Keep the record but detach and delete the attached file
3. Keep the file but remove the record

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:480]: I think re: my comment on the very top. This section can probably be collapsed into a warning that's at the very top of this page.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:32]: To clarify - your suggestion is that we combine record creation, file upload, and file association with the record into a single function?

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:173]: The reason I chose "file" / "association" over "attachment" / "attached" was because I didn't want to confuse customers regarding what was actually happening here. We are creating an association between the record and the file, which can be later updated, but there is otherwise no library logic that connects these two in any way (unlike the AppSync SDK, which was performing the upload), and I was concerned that "attachment" may imply that. My preference would be to stick with "file / association", if that works for you.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:32]: Into a single "section". Re: if in the same function, I think we should just develop an opinion that customers should generally follow. If we think diff functions are better then let's do that.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:173]: Cool. Let's then use association throughout.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:32]: Re: order of operations, it depends on the customer use-case. If, for instance, there is a situation where a created record will always have an initial file, then we should perform the upload, then the mutation. In a situation where the file may be added later (for instance, adding a profile picture after a profile has been created), then we would create the record, upload the file, then update the record. I'll call this out in the documentation to make it clear.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:150]:
```suggestion
2. Remove the record association and delete the file.
```

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:191]:
```suggestion
### Remove the record association and delete the file
```

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:202]:
```suggestion
});
```

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:201]:
```suggestion
  query: queries.getSong,
  variables: { id: currentSong.id },
```

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:210]:
```suggestion
  coverArtKey: null, // Set the file association to `null`
```

**Review by renebrandel — APPROVED** [review]

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:3]: @renebrandel @david-mcafee  - do you want to use the fact that you can add the "api" category first with the Cognito user pool as an Auth method here to streamline the number of CLI commands?

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:12]: I think we should be a bit more specific that you should find your schema .graphql file, and where it is at - and update it.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:24]: I would pull this up right after the CLI commands - I think it would flow better that way

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:28]: I am a little confused on this being here - what are you trying to prevent?

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (cwomack) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:18]: Should we clarify here the "how" something is set as optional? Something like: 
"Set as optional by leaving out the trailing '!' to allow adding file after initial create"

**Review by nadetastic — DISMISSED** [review]
Storage specific info/operations LGTM

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:3]: @abdallahshaban557 - in that case, we'd have to add instructions on adding auth. Telling them to run a single command with default configurations is simpler, IMO. @renebrandel?

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:12]: @abdallahshaban557 - I'm not absolutely opposed, but I was attempting to not re-explain items that are found elsewhere in the docs (such as in the `Getting Started` guides). Given that this isn't a separate tutorial, and is found in the context of our docs, I'm inclined to say no. However, it's a minor update, so I'm okay either way.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:18]: Because this isn't a standalone tutorial and is in the context of our other docs, I try to refrain from going into too much unrelated detail that would clutter the examples / explanations. If the customer needs more details on creating / modifying a schema, my assumption is that they will visit the relevant docs pages. Since there is a required `name` field above this, it also seems fairly straightforward by doing a quick visual comparison.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:46]: @cwomack - there are two ways to create a record with an associated file: 1) upload the Storage file, then create the API record with the file association, and 2) create the API record, upload the Storage file, then update the API record to create the association. I put the callout before the snippet because I would like to make it clear to the customer that there are two workflows, and I'd like to contain that guidance in one place. If we put this below the snippet, the customer may not clearly understand that there are two initial workflows.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (iartemiev) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:18]: Typo: `covert` instead of `cover`

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (iartemiev) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:24]:
```suggestion
To use files and attachments your application needs authorization credentials for reading and writing to Amazon S3. This is configured in the `amplify add auth` step.
```

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (iartemiev) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:107]:
```suggestion
  coverArtKey: result?.key,
```

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (iartemiev) [inline src/fragments/lib/graphqlapi/js/working-with-files.mdx:128]:
```suggestion
  coverArtKey: result?.key,
```

**Review by iartemiev — APPROVED** [review]
LGTM. Just a few nits

**Review by cwomack — APPROVED** [review]
LGTM

**Comment by david-mcafee** [comment]
> LGTM. Just a few nits

Thank you for catching these, Ivan!

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx]: Note: the wording here has been updated (see Ivan's suggestion), which is why this is marked as "outdated" - however, the placement is the same.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:12]: I would rather we mention it here IMO. It would make this section a lot clearer.

**Review by renebrandel — APPROVED** [review]

**Review by renebrandel — DISMISSED** [review]

**Review by renebrandel — APPROVED** [review]

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx]: LGTM

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (abdallahshaban557) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:36]: ok got it - thank you.

File: src/fragments/lib/graphqlapi/js/working-with-files.mdx
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx:20]: I should also add that the CLI prompts you to open the file for editing, so this may be redundant.

Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/working-with-files.mdx]: Added this in, marked as resolved

**Review by renebrandel — DISMISSED** [review]

**Review by nadetastic — APPROVED** [review]

**Review by abdallahshaban557 — APPROVED** [review]
LGTM!

**Review by renebrandel — APPROVED** [review]

**Review by renebrandel — APPROVED** [review]

**Review by nadetastic — APPROVED** [review]

**Review by katiegoines — APPROVED** [review]

**Review by cwomack — COMMENTED** [review]
Still LGTM (especially after it's merged 🚀)

**Pull request merged**
